great_expectations/experimental/datasources/data_asset/data_connector/data_connector.py

```diff
@@ -0,0 +1,190 @@
+from __future__ import annotations
+
+import logging
+import pathlib
+from typing import TYPE_CHECKING, Any, Dict, List, Optional
+
+from great_expectations.core.id_dict import BatchSpec
+
+if TYPE_CHECKING:
+    from great_expectations.core.batch import BatchDefinition
+    from great_expectations.experimental.datasources.interfaces import BatchRequest
+
+
+logger = logging.getLogger(__name__)
+
+
+# noinspection SpellCheckingInspection
+class DataConnector:
+    """The base class for all Data Connectors.
+
+    Data Connectors produce identifying information, called Batch Specs, that Execution Engines
+    can use to get individual batches of data. They add flexibility in how to obtain data
+    such as with time-based partitioning, downsampling, or other techniques appropriate
+    for the Datasource.
+
+    For example, a DataConnector could produce a SQL query that logically represents "rows in
+    the Events table with a timestamp on February 7, 2012," which an SqlAlchemy Datasource
+    could use to materialize a SqlAlchemy Dataset corresponding to that Batch of data and
+    ready for validation.
+
+    A Batch is a sample from a data asset, sliced according to a particular rule. For example,
+    an hourly slide of the Events table or “most recent Users records.” It is the primary
+    unit of validation in the Great Expectations Data Context. Batches include metadata that
+    identifies how they were constructed--the same Batch Spec assembled by the data connector.
+    While not every Datasource will enable re-fetching a specific batch of data, GX can store
+    snapshots of batches or store metadata from an external data version control system.
+
+    Args:
+        name: The name of the DataConnector instance
+        datasource_name: The name of the Datasource associated with this DataConnector instance
+        data_asset_name: The name of the DataAsset using this DataConnector instance
+    """
+
+    def __init__(
+        self,
+        name: str,
+        datasource_name: str,
+        data_asset_name: str,
+    ) -> None:
+        self._name: str = name
+
+        self._datasource_name: str = datasource_name
+        self._data_asset_name: str = data_asset_name
+
+        # This is a dictionary which maps data_references onto batch_requests.
+        self._data_references_cache: Dict = {}
```

Comment:
this variable is not needed here, and could benefit from some type hints

Reply:
@donaldheppner This is needed and used.  The line above initializes the cache between data references (e.g., file paths) and the `BatchDefinition` objects.  Caching paths can enables the listing of large directories on remote/cloud file systems (e.g., Web logs on S3) to be less frequent when we can find the references in the cache.  I updated the type hint to be more descriptive.  In this top-level `DataConnector` class, `_data_references_cache` is used only in `self_check()`, but its main use is in the `FilePathDataConnector` extension, where high-level path operations are implemented.  Thanks.